yt/yt/library/decimal/decimal.cpp

```diff
@@ -26,10 +144,138 @@ struct TDecimalTraits
+    static constexpr i256 Nan = {UInt32Max, UInt32Max, UInt32Max, UInt32Max, UInt32Max, UInt32Max, UInt32Max, Int32Max};
+    static constexpr i256 PlusInf = {UInt32Max - 1, UInt32Max, UInt32Max, UInt32Max, UInt32Max, UInt32Max, UInt32Max, Int32Max};
+    static constexpr i256 MinusInf = -PlusInf;
 };
 
+template <typename T>
+Y_FORCE_INLINE constexpr bool IsNegativeInteger(T value)
+{
+    static_assert(ValidDecimalUnderlyingInteger<T>);
+
+    if constexpr (std::is_same_v<T, i256>) {
+        return value.Parts[7] & (1u << 31);
+    } else {
+        return value < 0;
+    }
+}
+
+template <typename T>
+Y_FORCE_INLINE constexpr auto DecimalIntegerToUnsigned(T value)
+{
+    static_assert(ValidDecimalUnderlyingInteger<T>);
+
+    if constexpr (std::is_same_v<T, i256>) {
+        return ui256{value.Parts};
+    } else if constexpr (std::is_same_v<T, i128>) {
+        return ui128(value);
+    } else {
+        using TU = std::make_unsigned_t<T>;
+        return static_cast<TU>(value);
+    }
+}
+
+template <typename T>
+Y_FORCE_INLINE constexpr auto DecimalIntegerToSigned(T value)
+{
+    static_assert(ValidDecimalUnderlyingUnsignedInteger<T>);
+
+    if constexpr (std::is_same_v<T, ui256>) {
+        return i256{value.Parts};
+    } else if constexpr (std::is_same_v<T, ui128>) {
+        return i128(value);
+    } else {
+        using TU = std::make_signed_t<T>;
+        return static_cast<TU>(value);
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+//! The functions below are used for implementing conversion to/from text to binary decimal.
+//! We actually do not need any arithmetic operations beside v *= 10, v /= 10, addition and negation.
+//! so they are the only ones actually implemented for our custom i256/ui256.
+
+template <typename T>
+Y_FORCE_INLINE constexpr auto MSBFlip(T value)
+{
+    static_assert(ValidDecimalUnderlyingInteger<T>);
+
+    if constexpr (std::is_same_v<T, i256>) {
+        value.Parts[7] ^= (1u << 31);
+        return value;
+    } else {
+        constexpr auto one = DecimalIntegerToUnsigned(T{1});
+        // Bit operations are only valid with unsigned types.
+        return T(DecimalIntegerToUnsigned(value) ^ (one << (sizeof(T) * 8 - 1)));
+    }
+}
+
+template <typename T>
+Y_FORCE_INLINE constexpr ui32 GetNextDigit(T value, T* nextValue)
+{
+    static_assert(ValidDecimalUnderlyingUnsignedInteger<T>);
+
+    if constexpr (std::is_same_v<T, ui256>) {
+        ui64 remainder = 0;
+        for (int partIndex = 7; partIndex >= 0; --partIndex) {
+            // Everything should fit into long long since we are dividing by 10.
+            auto step = std::lldiv(value.Parts[partIndex] + (remainder << 32), 10);
+            value.Parts[partIndex] = step.quot;
+            remainder = step.rem;
+        }
+        *nextValue = value;
+        return remainder;
+    } else {
+        constexpr auto ten = T{10};
+        *nextValue = value / ten;
```

Comment:
Actually, you can also optimize this case with lldiv but who cares :)

Reply:
It is annoying to do because it is not a single overloaded function for all integral types, but rather a set of functions (ldiv, lldiv, etc). So yeah, too much effort.